accel/tcg/log_instr.c

```diff
@@ -0,0 +1,513 @@
+    /*     log->cv_buffer.val5 = cap_get_length64(cr); */
+    /* } */
+}
+
+static void emit_log_start(CPUArchState *env, target_ulong pc)
+{
+    if (qemu_loglevel_mask(CPU_LOG_INSTR)) {
+        qemu_log("Requested instruction logging @ " TARGET_FMT_lx
+                 " ASID %u\n", pc, cheri_get_asid(env));
+        qemu_log("User-mode only tracing enabled at " TARGET_FMT_lx
+                 ", ASID %u\n", pc, cheri_get_asid(env));
+    } else if (qemu_loglevel_mask(CPU_LOG_CVTRACE)) {
+        // TODO(am2419) Emit an event for instruction logging start
+    }
+}
+
+static void emit_log_stop(CPUArchState *env, target_ulong pc)
+{
+    if (qemu_loglevel_mask(CPU_LOG_INSTR)) {
+        if (qemu_loglevel_mask(CPU_LOG_USER_ONLY))
+            qemu_log("User-mode only tracing disabled at " TARGET_FMT_lx
+                     ", ASID %u\n", pc, cheri_get_asid(env));
+        else
+            qemu_log("Disabled instruction logging @ " TARGET_FMT_lx
+                     " ASID %u\n", pc, cheri_get_asid(env));
+    } else if (qemu_loglevel_mask(CPU_LOG_CVTRACE)) {
+        // TODO(am2419) Emit an event for instruction logging stop
+    }
+}
+
+static void reset_log_buffer(cpu_log_instr_info_t *log)
+{
+    g_array_remove_range(log->regs, 0, log->regs->len);
+    g_string_erase(log->txt_buffer, 0, -1);
+    log->force_drop = false;
+}
+
+void helper_qemu_log_instr_start(CPUArchState *env, target_ulong pc)
+{
+    cpu_log_instr_info_t *log = target_cpu_get_log(env);
+
+    qemu_log_instr_start(env, cl_default_trace_format, pc);
+    /* Skip this instruction commit */
+    log->force_drop = true;
+}
+
+void helper_qemu_log_instr_stop(CPUArchState *env, target_ulong pc)
+{
+    qemu_log_instr_stop(env, cl_default_trace_format, pc);
+}
+
+void helper_qemu_log_instr_user_start(CPUArchState *env, target_ulong pc)
+{
+    cpu_log_instr_info_t *log = target_cpu_get_log(env);
+
+    qemu_log_instr_start(env, cl_default_trace_format | CPU_LOG_USER_ONLY, pc);
+    /* Skip this instruction commit */
+    log->force_drop = true;
+}
+
+void helper_qemu_log_instr_user_stop(CPUArchState *env, target_ulong pc)
+{
+    qemu_log_instr_stop(env, cl_default_trace_format | CPU_LOG_USER_ONLY, pc);
+}
+
+void helper_qemu_log_instr_commit(CPUArchState *env)
+{
+    _qemu_log_instr_commit(env);
+}
+
+void helper_qemu_log_instr(CPUArchState *env, target_ulong pc)
+{
+    _qemu_log_instr_instr(env, pc);
+}
+
+void qemu_log_instr_init(CPUArchState *env)
+{
+    cpu_log_instr_info_t *log = target_cpu_get_log(env);
+
+    memset((void *)log, 0, sizeof(*log));
+    log->txt_buffer = g_string_new(NULL);
+    log->regs = g_array_new(FALSE, TRUE, sizeof(log_reginfo_t));
+    log->force_drop = true;
+}
+
+/*
+ * Callback invoked by tcg when flushing on state changed
+ * TODO(am2419): should we move the flush callback here? it seems to be
+ *  instr logging-related and only used for this.
+ */
+/*
+void flush_tcg_on_log_instr_chage(void);
```

Comment:
The only reason this ugly hack is needed is that log.c can't use tcg functions since it's also included in binaries such as qemu-img..